### PR TITLE
Allows source edits from Translation.io

### PR DESCRIPTION
**Pull request by MichaelHoste**
Hello @armandsar,

Here is a pull request to allow source edits from Translation.io (read here for more info: https://translation.io/blog/new-feature-copywriting)

The main idea is to change source text AND translations from a centralised place. It allows copywriting from the marketing/management team without annoying developers.

Also, I added [a snippet](https://gist.github.com/lithrel/a224edb1ed2975992c73) to pretty print translation files. So they are easier to read.
 
If you have any comment on the produced code, please tell me. As you know, I'm not very confortable with PHP since that's not my language of choice.

I think this package is already mature enough to be distributed to the Laravel community. I intend to create https://laravel.translation.io in the following month(s) and start to look for beta-testers.

So you know, I think the package still need to deal with [special translation management of Laravel >5 with sub-folders and keys like "this/is/repository.and.here.are.keys"](https://laravel.io/forum/02-23-2015-localization-load-files-from-subdirectories-at-resourceslanglocale). But I doubt I lot of users use that feature so it can certainly wait.

I added my name as author on the composer file, I hope you don't mind.

Oh, and Happy Christmas ! 🎅 🎄

### CODE REVIEW

**Comment by armandsar** [comment]
@MichaelHoste nice feature, thanks. I might have some comments, will take a deeper look after holidays.
One thing I noticed worth doing - creating SourceEditSync class, Sync is a bit too big now, I think.
By the way, do you feel adventurous enough to try writing tests for this? If not thats fine, I will might do it later.

https://laravel.translation.io  - sounds great! 

Merry Christmas!

**Comment by MichaelHoste** [comment]
You're right, I'll try to create a SourceEditSync class. First I didn't do it because I wanted to reuse the code to build of the request. But it would certainly be cleaner. Also, it would make the VCR test easier to do.

I'll try to make some tests (and fix the failed ones) after the holidays.

Thank you for your quick feedback!

**Comment by MichaelHoste** [comment]
Hi @armandsar,

I improved the code by extracting SourceEditSync like you said. I did several other improvements too and fixed the CodeClimate warnings. I hope you'll agree with all this.

As for the tests, there is still one failed test because of this line: https://github.com/MichaelHoste/laravel-translationio/blob/master/tests/Integration/SyncTest.php#L26

I don't know how to directly call SyncService class instead of calling the sync command with artisan (that is doing 2 requests now). It's certainly not too difficult but I need some help with PHP here :-/

If you can help me do that, I'll try to make some more tests for SourceEditSync (+ PrettyVarExport and SourceSaver if I have time).

**Comment by armandsar** [comment]
@MichaelHoste I have added source edit request (with blank response) to vcr, that should pass the test. I was thinking of doing two tests, but given that its an integration test for console command maybe mocking any of the service might be wrong

File: src/PrettyVarExport.php [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+class PrettyVarExport
+{
+    /**
+     * @var Application
+     */
+    private $application;
+
+    public function __construct(Application $application)
+    {
+        $this->application = $application;
```
Comment (armandsar) [inline src/PrettyVarExport.php]: I dont see this used anywhere

File: src/Service/SourceEditSync.php [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+            return 0;
+        }
+        else {
+            return $json['timestamp'];
+        }
+    }
+
+    private function throwErrorIfConflictInMetadata($metadataContent)
+    {
+      if (strpos($metadataContent, '>>>>') !== false || strpos($metadataContent, '<<<<') !== false) {
```
Comment (armandsar) [inline src/Service/SourceEditSync.php]: You can use 
str_contains($string, ['>>>>', '<<<<'])

File: src/Service/SourceEditSync.php [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+          $metadataFilePath = $this->metadataFilePath();
+          exit("[Error] " . $metadataFilePath . " file is corrupted and seems to have unresolved versioning conflicts. Please resolve them and try again.");
```
Comment (armandsar) [inline src/Service/SourceEditSync.php]: Given the name of method, we could actually do raise exception here

File: src/SourceSaver.php [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    }
+
+    private function keys($key)
+    {
+        $keyParts = explode('.', $key);
+        array_shift($keyParts); // remove group part
+        return $keyParts;
+    }
+
+    private function replaceValueInTranslations(&$translations, $sourceEdit)
```
Comment (armandsar) [inline src/SourceSaver.php]: I dont really like the use of "reference" here, but as long as this works no big deal, we can return back to this and refactor when we have tests covering this part

**Comment by MichaelHoste** [comment]
@armandsar I made the changes based on your feedback, and I added a complete sourceEdit test by completing the blank VCR. I guess it should be alright now.

Travis failed because of a timeout on one of its instance.

**Pull request merged**